containers: T3662: Repair system damage during upgrade

jack9603301 opened the pull request:
<!-- All PR should follow this template to allow a clean and transparent review -->
<!-- Text placed between these delimiters is considered a commend and is not rendered -->

## Change Summary
<!--- Provide a general summary of your changes in the Title above -->
Repair system damage during upgrade

## Types of changes
<!---
What types of changes does your code introduce? Put an 'x' in all the boxes that apply.
NOTE: Markdown requires no leading or trailing whitespace inside the [ ] for checking
the box, please use [x]
-->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)
<!-- All submitted PRs must be linked to a Task on Phabricator. -->
* https://phabricator.vyos.net/T3662

## Component(s) name
<!-- A rather incomplete list of components: ethernet, wireguard, bgp, mpls, ldp, l2tp, dhcp ... -->
containers

## Proposed changes
<!--- Describe your changes in detail -->
Because the new vyos system may not have the OCI runtime `runc` used by podman by default, but there is a crun, resulting in the container configuration can not be used normally. We need to point to the `crun` through the soft link when the system can not find the `runc`.

This patch will fix the above problem.

## How to test
<!---
Please describe in detail how you tested your changes. Include details of your testing
environment, and the tests you ran. When pasting configs, logs, shell output, backtraces,
and other large chunks of text, surround this text with triple backtics
```
like this
```
-->

When upgrading a vyos system containing container configuration, the configuration will be destroyed, such as the following configuration:

```
set interfaces ethernet eth0 address dhcp
set interfaces ethernet eth0 ipv6 address autoconf
set system name-service 114.114.114.114
set container name nginx image nginx
set container name nginx allow-host-networks
commit
save
exit
```

After upgrading to vyos compiled by this branch, Restart after the first upgrade:
![图片](https://user-images.githubusercontent.com/9047180/124386710-3ac1eb80-dd0e-11eb-8008-3bc70c5231eb.png)



## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- The entire development process is outlined here: https://docs.vyos.io/en/latest/contributing/development.html -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [ ] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

jack9603301 committed:
containers: T3662: Repair system damage during upgrade

Because the new vyos system may not have the OCI runtime
`runc` used by podman by default, but there is a crun,
resulting in the container configuration can not be used normally.

This patch will fix the above problem.

sever-sever commented:
If you place storage to **/var** how it will copy container images/contents/metadata between updates?

jack9603301 commented:
> If you place storage to **/var** how it will copy container images/contents/metadata between updates?

Volume map can be used to map data area persistence to `/config`, which is in line with the best practice of container microservice technology

Only valid persistent data needs to be copied, and no other data and metadata need to be copied. When the new system is configured for the first time, the image will be pulled directly from the docker. The container itself is suitable for stateless services. If the persistent data mapping is valid and undamaged, it should be able to be used by the newly pulled container image.

sever-sever commented:
Check this behavior.
Update system image, Disable interfaces or remove the default route. To simulate not internet connection.
After that, reboot the router to the updated version.

jack9603301 commented:
At the same time, I increased the priority, which will make the container configuration execute after the network configuration.

After the system is upgraded and rebooted for the first time, all configurations will be migrated to the new system. Due to the first pull, it is necessary to connect to the Internet. This is a normal behavior. If enterprises need to pull from the intranet, they can easily establish a private image server such as harbor. Podman can pull from the intranet server instead of the public network

Therefore, this situation generally does not need to be considered

jack9603301 commented:
In the case of no network, the network configuration is normal due to priority, but the service of container (priority > = 1200) cannot be configured

![图片](https://user-images.githubusercontent.com/9047180/124388223-89728400-dd14-11eb-99bc-613cebac0442.png)
![图片](https://user-images.githubusercontent.com/9047180/124388226-8d9ea180-dd14-11eb-8d13-6944760fbbaa.png)
![图片](https://user-images.githubusercontent.com/9047180/124388233-92fbec00-dd14-11eb-8a53-85ca4aea8d5f.png)

Before applying this patch, when a user using container configuration upgrades to a new image, the system will not be able to configure the network (lack of priority) due to container configuration failure. At the same time, the container configuration appears to have been corrupted, so you must delete and reconfigure the container

c-po commented on data/templates/containers/storage.tmpl:
```diff
@@ -2,4 +2,4 @@
 [storage]
   driver = "vfs"
-  graphroot = "/config/containers/storage"
+  graphroot = "/var/containers/storage"
```
This will not survive an image upgrade

jack9603301 replied:
For containers that support volume mapping, it does not need to exist

jack9603301 replied:
After the upgrade is started for the first time, the image will be pulled again, and the persistent storage can be used normally through volume mapping (yes, as long as the user sets the container correctly, this is in line with the best practice of container and microservice Technology)

sever-sever commented:
As for me, it needs to find out the root cause.
Maybe needs to add some additional keys/values for section **[storage]** or use another **driver**.

jack9603301 commented:
@c-po @sever-sever Sorry, I didn't mean anything else. I just think that this problem can be solved according to normal container best practice. If you persist metadata permanently, I have a feeling that volume persistence mapping has lost its role

Persistent volume mapping has been implemented here:

https://github.com/vyos/vyos-1x/pull/818

Example, Use a similar command line:

```
set container name adguardhome allow-host-networks
set container name adguardhome description 'AdGuardHome'
set container name adguardhome image 'adguard/adguardhome'
set container name adguardhome volume conf destination '/opt/adguardhome/conf'
set container name adguardhome volume conf source '/config/adguardhome/conf'
set container name adguardhome volume work destination '/opt/adguardhome/work'
set container name adguardhome volume work source '/config/adguardhome/work'
set container name pac allow-host-networks
set container name pac description 'PAC-NGINX'
set container name pac image 'nginx'
set container name pac volume nginx-conf destination '/etc/nginx/conf.d'
set container name pac volume nginx-conf source '/config/pac/vhost'
set container name pac volume pac destination '/www/root/default'
set container name pac volume pac source '/config/pac/www'
```

For the home network, it can be pulled from the external network. When upgrading the system in the enterprise and it needs to be pulled from the internal network, a dedicated server such as Harhor can be used to store and pull the image.

:)

https://goharbor.io/

sever-sever commented:
Most likely it will work.
```
root@r1-roll:/home/vyos# podman --runtime /usr/bin/crun run --net host --rm fedora echo it works
Error: OCI runtime error: writing file `/sys/fs/cgroup/cgroup.subtree_control`: Invalid argument
root@r1-roll:/home/vyos# 

root@r1-roll:/home/vyos# podman --runtime /usr/bin/runc run --net host --rm fedora echo it works
it works
root@r1-roll:/home/vyos#

```

sever-sever reviewed (commented):
Most likely it will work.
```
root@r1-roll:/home/vyos# podman --runtime /usr/bin/crun run --net host --rm fedora echo it works
Error: OCI runtime error: writing file `/sys/fs/cgroup/cgroup.subtree_control`: Invalid argument
root@r1-roll:/home/vyos# 

root@r1-roll:/home/vyos# podman --runtime /usr/bin/runc run --net host --rm fedora echo it works
it works
root@r1-roll:/home/vyos#

```
But in my test vm by default, after install runc it doesn't use runc
```
root@r1-roll:/home/vyos# podman  run --net host --rm fedora echo it works
Error: OCI runtime error: writing file `/sys/fs/cgroup/cgroup.subtree_control`: Invalid argument
root@r1-roll:/home/vyos# 

```
ociRuntime still **crun**
```
root@r1-roll:/home/vyos# podman info | grep ociRuntime -A 8
  ociRuntime:
    name: crun
    package: 'crun: /usr/bin/crun'
    path: /usr/bin/crun
    version: |-
      crun version 0.17
      commit: 0e9229ae34caaebcb86f1fde18de3acaf18c6d9a
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +YAJL

```

jack9603301 commented:
@sever-sever 
![图片](https://user-images.githubusercontent.com/9047180/124559273-21d04c00-de6e-11eb-9f4b-c24511cac600.png)
Strangely enough, I can't reproduce this mistake

jack9603301 committed:
containers: T3662: Forces the specified OCI runtime to be used

In some cases, using `crun` seems to lead to run-time exceptions,
so we have to add parameters related to the mandatory use of `runc`

sever-sever commented:
The root cause in changing in **cgroups**/cgroup2 with disabled option **--cgroups disabled** I don't see any errors, even with **crun**
```
root@r1-roll:/home/vyos# podman  run --cgroups disabled --net host --rm fedora echo it works
it works
root@r1-roll:/home/vyos#
```

Debian 10 use cgroup v1
```
vyos@r1-roll:~$ sudo mount | grep cgroup
tmpfs on /sys/fs/cgroup type tmpfs (ro,nosuid,nodev,noexec,mode=755)
cgroup2 on /sys/fs/cgroup/unified type cgroup2 (rw,nosuid,nodev,noexec,relatime,nsdelegate)
cgroup on /sys/fs/cgroup/systemd type cgroup (rw,nosuid,nodev,noexec,relatime,xattr,name=systemd)
cgroup on /sys/fs/cgroup/net_cls,net_prio type cgroup (rw,nosuid,nodev,noexec,relatime,net_cls,net_prio)
cgroup on /sys/fs/cgroup/cpuset type cgroup (rw,nosuid,nodev,noexec,relatime,cpuset)
cgroup on /sys/fs/cgroup/cpu,cpuacct type cgroup (rw,nosuid,nodev,noexec,relatime,cpu,cpuacct)
cgroup on /sys/fs/cgroup/blkio type cgroup (rw,nosuid,nodev,noexec,relatime,blkio)
cgroup on /sys/fs/cgroup/devices type cgroup (rw,nosuid,nodev,noexec,relatime,devices)
cgroup on /sys/fs/cgroup/pids type cgroup (rw,nosuid,nodev,noexec,relatime,pids)
cgroup on /sys/fs/cgroup/memory type cgroup (rw,nosuid,nodev,noexec,relatime,memory)
cgroup on /sys/fs/cgroup/perf_event type cgroup (rw,nosuid,nodev,noexec,relatime,perf_event)
cgroup on /sys/fs/cgroup/freezer type cgroup (rw,nosuid,nodev,noexec,relatime,freezer)
vyos@r1-roll:~$
root@r1-roll:/home/vyos# podman info
host:
  arch: amd64
  buildahVersion: 1.19.6
  cgroupManager: systemd
  cgroupVersion: v1
```
Debian 11 use cgroup v2
```
vyos@r1-roll:~$ sudo mount | grep cgroup
cgroup2 on /sys/fs/cgroup type cgroup2 (rw,nosuid,nodev,noexec,relatime,nsdelegate,memory_recursiveprot)
vyos@r1-roll:~$ 
root@r1-roll:/home/vyos# podman info
host:
  arch: amd64
  buildahVersion: 1.19.6
  cgroupManager: systemd
  cgroupVersion: v2
```

sever-sever commented:
In the latest rolling release all works fine without any changes
```
vyos@r1-roll:~$ show version 

Version:          VyOS 1.4-rolling-202107010537
Release Train:    sagitta

vyos@r1-roll:~$ sudo podman  run --net host --rm fedora echo it works
it works
vyos@r1-roll:~$ 
```

jack9603301 commented:
@sever-sever 
Why do you have problems after adding runc?

jack9603301 commented:
note: Record the process of upgrading from 1.4-rolling-202107010537 to 1.4-rolling-202107122017

After the first startup, the container fails to start. Restart the operating system and start normally.

Wait for the next test to determine if this problem is resolved!

jack9603301 commented:
![图片](https://user-images.githubusercontent.com/9047180/126420223-846cd1f9-ee19-4e90-b749-a1730d4052b9.png)

Unfortunately, it seems that the same problem still happened. Although it seems that the container can be started normally after restart due to priority adjustment and some factors, it is not a normal operation result!

c-po commented:
I do not understand why adding the `runc` dependency should fix this issue. I just upgraded to the latest 1.4 rolling release and the container was starting up as expected with all the data present.

I propose to close this PR.

jack9603301 commented:
@c-po I will re evaluate whether the problem disappears. If it does, the PR will be closed

jack9603301 commented:
![图片](https://user-images.githubusercontent.com/9047180/130771258-cb5cde10-c8cc-4d28-a8be-41940b9671af.png)
@c-po This problem still exists and needs to be normal after a second full restart, but one of the errors may be caught

c-po commented:
I have just tried this PR out of curiosity but as expected it does not work, `runc` is like `podman` a program to run containers and used by e.g. Docker.

The issue is that podman does not find the container image on first invocation on the first boot but on any other subsequent boot.

c-po commented:
Closing this as this does not solve the root cause. The root cause "seems" to be the fact that on system bootup and container initialisation there might still not be a full propagated routing table so podman can reach out to the registry.

This might require a conceptual change in how Containers are presented.
